pkg/provisioning/provisioning.go

```diff
@@ -4,136 +4,35 @@ import (
-	// hardware details struct as part of a second pass.
-	if host.Status.HardwareDetails == nil {
-		reqLogger.Info("continuing inspection by setting details")
-		host.Status.HardwareDetails = &metalkubev1alpha1.HardwareDetails{}
-		return true, nil
-	}
-
-	return false, nil
-}
-
-// Deprovision prepares the host to be removed from the cluster. It
-// may be called multiple times, and should return true for its dirty
-// flag until the deprovisioning operation is completed.
-func (p *Provisioner) Deprovision(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error) {
-	reqLogger := log.WithValues("host", host.Name)
-	reqLogger.Info("ensuring host is removed")
-
-	if host.Status.HardwareDetails != nil {
-		reqLogger.Info("clearing hardware details")
-		host.Status.HardwareDetails = nil
-		return true, nil
-	}
-
-	if host.Status.Provisioning.ID != "" {
-		reqLogger.Info("clearing provisioning id")
-		host.Status.Provisioning.ID = ""
-		return true, nil
-	}
-
-	return false, nil
-}
-
-// PowerOn ensures the server is powered on independently of any image
-// provisioning operation.
-func (p *Provisioner) PowerOn(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error) {
-	reqLogger := log.WithValues("host", host.Name)
-	reqLogger.Info("ensuring host is powered on")
-
-	if dirty, err = p.ensureExists(host); err != nil {
-		return dirty, err
-	}
-
-	if host.Status.Provisioning.State != "powered on" {
-		host.Status.Provisioning.State = "powered on"
-		return true, nil
-	}
-
-	return false, nil
-}
-
-// PowerOff ensures the server is powered off independently of any image
-// provisioning operation.
-func (p *Provisioner) PowerOff(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error) {
-	reqLogger := log.WithValues("host", host.Name)
-	reqLogger.Info("ensuring host is powered off")
-
-	if dirty, err = p.ensureExists(host); err != nil {
-		return dirty, err
-	}
-
-	if host.Status.Provisioning.State != "powered off" {
-		host.Status.Provisioning.State = "powered off"
-		return true, nil
-	}
-
-	return false, nil
+type Provisioner interface {
+	// ValidateManagementAccess tests the connection information for the
+	// host to verify that the location and credentials work.
+	ValidateManagementAccess(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error)
+
+	// InspectHardware updates the HardwareDetails field of the host with
+	// details of devices discovered on the hardware. It may be called
+	// multiple times, and should return true for its dirty flag until the
+	// inspection is completed.
+	InspectHardware(host *metalkubev1alpha1.BareMetalHost) (dirty bool, err error)
+
+	// Deprovision prepares the host to be removed from the cluster. It
+	// may be called multiple times, and should return true for its dirty
+	// flag until the deprovisioning operation is completed.
+	Deprovision(host *metalkubev1alpha1.BareMetalHost) (dirty bool, requeueDelay time.Duration, err error)
```

Comment:
I imagine we'll add a Provision(), and just haven't gotten there yet?

Reply:
Yeah, that'll come soon.